File: Longest Consecutive Sequence.py
# https://leetcode.com/problems/longest-consecutive-sequence/

import sys

class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        
        # If list is empty, return 0
        if len(nums) == 0:
            return 0
        
        # If list only has 1 element, return 1
        if len(nums) == 1:
            return 1
        
        # Add largest number at the end of the list
        maxNum = sys.maxsize
        nums.append(maxNum)
        
        # Sort list
        nums.sort()
        
        # Create sequence variables
        # NOTE: Count 1st digit as start of the sequence, so its 1
        longestSeq = 0
        tmpSeq = 1
        
        # Iterate list
        for idx in range(1, len(nums)):
            
            # If next digit is maxNum, then we have reached the end of the list
            if nums[idx] == maxNum:
                break
            
            # If difference between this and previous digit is greater than 1, then update longestSeq and reset counter
            if nums[idx] - nums[idx - 1] > 1:
                longestSeq = max(longestSeq, tmpSeq)
                tmpSeq = 1
                
            # If this and previous number are the same, don't increment sequence counter
            elif nums[idx] - nums[idx - 1] == 0:
                continue
            
            else:
                tmpSeq += 1
                
        # Recalculate max value in case sequence ended when maxNum was encountered
        return max(longestSeq, tmpSeq)

File: Longest Consecutive Sequence.java
// https://leetcode.com/problems/longest-consecutive-sequence/
// https://youtu.be/P6RZZMu_maU?si=iS0FHXzS7ax7ZA3p

/*
Time Complexity: O(n) where n = length of input array.
Space Complexity: O(n) where n = length of set.
*/

class Solution {

    private Set<Integer> generateSetFromArray(int[] nums) {
        Set<Integer> uniqueNums = new HashSet<>();

        for (int number: nums) {
            uniqueNums.add(Integer.valueOf(number));
        }

        return uniqueNums;
    }

    private int calculateLongestSubsequence(Set<Integer> uniqueNums) {
        
        int longestLength = 0;

        // E.g., Assume curNumber is 10
        for (Integer curNumber: uniqueNums) {

            // Check that 9 doesn't exist inside set.
            // Why? Because we don't want to count a sequence twice. We should just go through the sequence in its entirety.
            if (!uniqueNums.contains(curNumber - 1)) {

                // From this idx, calculate longest subsequence
                // Does 10 + 1 exist? Does 10 + 2 exist? Etc.)
                int curLength = 0;
                while (uniqueNums.contains(curNumber + curLength)) {
                    curLength++;
                }

                // Update to the largest number
                longestLength = Math.max(longestLength, curLength);
            }
        }

        return longestLength;
    }

    public int longestConsecutive(int[] nums) {
        
        if (nums.length <= 1) {
            return nums.length;
        }

        Set<Integer> uniqueNums = generateSetFromArray(nums);

        int longestLength = calculateLongestSubsequence(uniqueNums);
        return longestLength;
    }
}
